lib/grammar-error.js

```diff
@@ -2,10 +2,29 @@
 
 // Thrown when the grammar contains an error.
 class GrammarError extends Error {
-  constructor(message, location) {
+  constructor(message, location, referenceLocation) {
     super(message);
     this.name = "GrammarError";
     this.location = location;
+    this.referenceLocation = referenceLocation;
+  }
+
+  toString() {
+    let str = super.toString();
+    if (this.location) {
+      const filename = (typeof this.location.source === "string")
+        ? `"${this.location.source}" `
+        : "";
+      str += `\nat ${filename}line ${this.location.start.line}, column ${this.location.start.column}`;
+    }
+    if (this.referenceLocation) {
+      const filename = (typeof this.referenceLocation.source === "string")
+        ? `"${this.referenceLocation.source}" `
+        : "";
+      str += `\nfrom ${filename}line ${this.referenceLocation.start.line}, column ${this.referenceLocation.start.column}`;
+    }
+
+    return str;
```

Comment:
Not necessary to follow my general suggestion above, because it will require access to the original string. Just enumerate errors with `"note: <message><location>"` will be enough. But if you wish, you can implement the Rust format if source grammar string was given :smiley:

Reply:
I'll make that a separate method.